tardis/visualization/widgets/custom_abundance.py

```diff
@@ -0,0 +1,593 @@
+"""Class to create and display Custom Abundance Widget."""
+import os
+import math
+import numpy as np
+import pandas as pd
+import ipywidgets as ipw
+import plotly.graph_objects as go
+from astropy import units as u
+from pyne import nucname
+
+from tardis.util.base import quantity_linspace
+from tardis.io.config_reader import Configuration
+from tardis.model import Radial1DModel
+from tardis.model.density import calculate_power_law_density, calculate_exponential_density
+from tardis.io.config_validator import validate_dict
+from tardis.io.parsers.csvy import load_csvy
+from tardis.io.model_reader import (
+    read_uniform_abundances,
+    parse_csv_abundances,
+)
+from tardis.util.base import (
+    atomic_number2element_symbol,
+    quantity_linspace
+)
+
+import asyncio
+
+class Timer:
+    """
+    Cited from https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html
+    """
+    def __init__(self, timeout, callback):
+        self._timeout = timeout
+        self._callback = callback
+
+    async def _job(self):
+        await asyncio.sleep(self._timeout)
+        self._callback()
+
+    def start(self):
+        self._task = asyncio.ensure_future(self._job())
+
+    def cancel(self):
+        self._task.cancel()
+
+def debounce(wait):
+    """ 
+    Decorator that will postpone a function's
+    execution until after `wait` seconds
+    have elapsed since the last time it was invoked. 
+    Cited from https://ipywidgets.readthedocs.io/en/latest/examples/Widget%20Events.html
+    """
+    def decorator(fn):
+        timer = None
+        def debounced(*args, **kwargs):
+            nonlocal timer
+            def call_it():
+                fn(*args, **kwargs)
+            if timer is not None:
+                timer.cancel()
+            timer = Timer(wait, call_it)
+            timer.start()
+        return debounced
+    return decorator
+
+class CustomAbundanceWidget:
```

Comment:
Brief docstrings for the `@properties` would be useful, particularly for the less immediately obvious ones.